kernel/src/transaction.rs

```diff
@@ -1,120 +1,260 @@
+        match json_handler.write_json_file(&commit_path.location, Box::new(actions), false) {
+            Ok(()) => Ok(CommitResult::Committed(commit_version)),
+            Err(crate::error::Error::ObjectStore(object_store::Error::AlreadyExists {
+                ..
+            })) => Ok(CommitResult::Conflict(self, commit_version)),
+            Err(e) => Err(e),
+        }
     }
 
-    /// Scan the Delta Log for the latest transaction entry of an application
-    pub fn application_transaction(
-        &self,
-        engine: &dyn Engine,
-        application_id: &str,
-    ) -> DeltaResult<Option<Transaction>> {
-        let mut transactions = self.scan_application_transactions(engine, Some(application_id))?;
-        Ok(transactions.remove(application_id))
+    /// Add commit info to the transaction. This is commit-wide metadata that is written as the
+    /// first action in the commit. Note it is required in order to commit. If the engine does not
+    /// require any commit info, pass an empty `EngineData`.
+    pub fn commit_info(&mut self, commit_info: Box<dyn EngineData>, schema: Schema) {
+        self.commit_info = Some(EngineCommitInfo {
+            data: commit_info.into(),
+            schema: schema.into(),
+        });
     }
+}
+
+/// Result after committing a transaction. If 'committed', the version is the new version written
+/// to the log. If 'conflict', the transaction is returned so the caller can resolve the conflict
+/// (along with the version which conflicted).
+pub enum CommitResult {
+    /// The transaction was successfully committed at the version.
+    Committed(crate::Version),
+    /// The transaction conflicted with an existing version (at the version given).
+    Conflict(Transaction, crate::Version),
+}
+
+// given the engine's commit info (data and schema as [EngineCommitInfo]) we want to create both:
+// (1) the commitInfo action to commit (and append more actions to) and
+// (2) the schema for the actions to commit (this is determined by the engine's commit info schema)
+fn generate_commit_info(
+    engine: &dyn Engine,
+    engine_commit_info: &EngineCommitInfo,
+) -> DeltaResult<(Box<dyn EngineData>, SchemaRef)> {
+    use crate::actions::{
+        ADD_NAME, COMMIT_INFO_NAME, METADATA_NAME, PROTOCOL_NAME, REMOVE_NAME, TRANSACTION_NAME,
+    };
 
-    /// Scan the Delta Log to obtain the latest transaction for all applications
-    pub fn application_transactions(&self, engine: &dyn Engine) -> DeltaResult<TransactionMap> {
-        self.scan_application_transactions(engine, None)
+    if engine_commit_info.data.length() != 1 {
+        return Err(crate::error::Error::InvalidCommitInfo(format!(
+            "Engine commit info should have exactly one row, found {}",
+            engine_commit_info.data.length()
+        )));
     }
+
+    let mut action_fields = get_log_schema().fields().collect::<Vec<_>>();
+    let commit_info_field = action_fields
```

Comment:
the naming is pretty confusing here. You have a `commit_info_field` and then from that you get `commit_info_fields`. Maybe call this `commit_info_struct_field` or something like that?

Reply:
fixed